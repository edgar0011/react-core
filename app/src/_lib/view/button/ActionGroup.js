// @flow
import React, { PureComponent } from 'react'
import type { Element, ComponentType, Node } from 'react'
import type { HeadlessRender } from '../common/HeadlessComponentBase'

type ComponentWithLabel = {
  label: string,
  component?: ComponentType<any>
}
type ActionGroupProps = {
  data: Array<ComponentWithLabel>,
  preData: Array<ComponentWithLabel>,
  postData: Array<ComponentWithLabel>,
  resolveButton: (props: ?Object) => ComponentType<any>,
  justify?: string
} & HeadlessRender

type PassProps = {
  data: Array<ComponentWithLabel>,
  preData: Array<ComponentWithLabel>,
  postData: Array<ComponentWithLabel>,
  justify?: string,
  buttons: Array<Element<any>>
}
/**
 * @class
 * @name ActionGroup
 * @classdesc Displays row of buttons typically within a form.
 * @extends PureComponent
 *
 * @prop {function} render Render function must be specified unless RenderComponent is specified.
 * @prop {ComponentType} RenderComponent Render component must be specified unless render function is specified.
 * @prop {Array<{ label: string }>} data Buttons configuration data.
 * @prop {Array<{ label: string, Component: any }>} preData List od items to be displayed before buttons.
 * @prop {Array<{ label: string, Component: any }>} postData List od items to be displayed after buttons.
 * @prop {function} resolveButton Button type resolver function.
 * @prop {string} justify View variations modifier.
 */
export default class ActionGroup extends PureComponent<ActionGroupProps> {

  constructor(props: ActionGroupProps) {
    super(props)
  }

  /**
   * Rendering nodes with appropriate button type render component.
   * @memberOf ActionGroup
   * @method ActionGroup#renderNodes
   * @param {Array<*>} buttonsData Buttons data.
   * @returns {Array<*>}
   */
  renderNodes = (buttonsData: Array<ComponentWithLabel>): Array<Element<any>> =>
    buttonsData.map(({ label, ...buttonProps }: { label: string, ...any}) => {
      const { resolveButton }: ActionGroupProps = this.props
      const Component: ComponentType<any> = resolveButton(buttonProps)
      const key: string = `${Component.toString()}-${label || buttonProps.title}`
      return <Component key={key} {...buttonProps} label={label} >{label}</Component>
    })

  /**
   * Component render function.
   * @memberOf ActionGroup
   * @method ActionGroup#render
   * @returns {*}
   */
  render() {
    const { data, render, RenderComponent, resolveButton, ...props }: ActionGroupProps = this.props
    const buttons: ?Array<Element<any>> = data && data.length ? this.renderNodes(data) : null
    if (!buttons) {
      return null
    }
    const passProps: PassProps = { buttons, data, ...props }
    let rendered: Node
    if (render) {
      rendered = render(passProps)
    } else {
      rendered = RenderComponent ? <RenderComponent {...passProps} /> : null
    }
    return rendered
  }
}
