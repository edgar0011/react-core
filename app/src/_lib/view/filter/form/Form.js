// @flow
import React, { PureComponent } from 'react'
import type { Node, Element, ComponentType } from 'react'
import type { HeadlessRender } from '../../common/HeadlessComponentBase'

export type FormComponents = {
  FormComponent: ComponentType<any>,
  FormFieldComponent: ComponentType<any>,
  ActionGroupComponent: ComponentType<any>
}

export type FlexLayout = {
  direction?: string,
  align?: string,
  self?: string,
  justify?: string,
  size?: number | string,
  padding?: number | string,
  margin?: number | string,
  wrap?: boolean,
  width?: string,
  minWidth?: string,
  maxWidth?: string
}

export type FlexFieldLayout = FlexLayout & {
  minWidth?: string
}

export type Field = {
  name: string,
  label?: string,
  Component: ComponentType<any>,
  value: any,
  valueDecorator?: (value: any, field: Field) => any,
  onChange?: (name: string, newValue: any) => void,
  render?: Function
}

export type Error = {
  code: string,
  message?: string
}

export type Errors = Array<Error> | any

export type RenderFieldProps = {
  field: Field,
  flexFieldLayout: FlexFieldLayout,
  fieldValue: any,
  components: FormComponents,
  fields: Array<Field>
}

export type FormProps = {
  fields: Array<Field>,
  errors?: Errors,
  flexFieldLayout: FlexFieldLayout,
  components: FormComponents,
  renderFieldMethod: (props: RenderFieldProps) => Array<Element<any>>
} & HeadlessRender

// aka FormBody
class Form extends PureComponent<FormProps> {
  props: FormProps

  render() {
    const {
      fields, flexFieldLayout, errors, render, RenderComponent, renderFieldMethod, components, ...formProps
    }: FormProps = this.props

    const hasErrors = !!(errors && (errors.length || Object.values(errors).length))

    const fieldNodes =
      fields.map(field => renderFieldMethod({ field, flexFieldLayout, fieldValue: field.value, components, fields }))

    const passProps = {
      fieldNodes,
      hasErrors,
      errors,
      components,
      ...formProps
    }

    let rendered: Node
    if (render) {
      rendered = render(passProps)
    } else {
      rendered = RenderComponent ? <RenderComponent {...passProps} /> : null
    }
    return rendered
  }
}

export default Form
