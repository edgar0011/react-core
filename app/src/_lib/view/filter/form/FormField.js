// @flow
import React, { PureComponent } from 'react'
import type { Node } from 'react'
import type { HeadlessRender } from '../../common/HeadlessComponentBase'

type FormFieldProps = {

} & HeadlessRender

class FormField extends PureComponent<FormFieldProps> {

  props: FormFieldProps

  render() {
    const { render, RenderComponent, ...props }: FormFieldProps = this.props

    let rendered: Node
    if (render) {
      rendered = render(props)
    } else {
      rendered = RenderComponent ? <RenderComponent {...props} /> : null
    }
    return rendered
  }
}

export default FormField
