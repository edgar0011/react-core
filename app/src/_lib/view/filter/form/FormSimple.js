// @flow
import React, { PureComponent } from 'react'
import { memoize } from 'lodash'

import type { Field, FlexFieldLayout, FormProps, FormComponents, FlexLayout, Errors } from './Form'

type FormSimpleProps = {
  fields: Array<Field>,
  errors: Errors,
  buttonLabels?: {
    reset?: string,
    submit?: string
  },
  formName: string,
  onChangeHandler?: Function,
  onBlurHandler?: Function,
  onFocusHandler?: Function,
  onResetHandler?: Function,
  onSubmitHandler: Function,
  submitDisabled?: boolean,
  showReset?: boolean,
  resetable?: boolean,
  showErrors?: boolean,
  flexLayout: FlexLayout,
  flexFieldLayout: FlexFieldLayout,
  translations?: {
    reset?: string,
    submit?: string
  },
  renderButtons?: Function,
  renderForm?: Function,
  renderField?: Function,
  components: FormComponents
}

type RenderButtons = {
  submitDisabled?: boolean,
  showReset?: boolean,
  resetable?: boolean,
  labels: {
    reset?: string,
    submit?: string
  },
  errors: Errors,
  components: FormComponents
}

type RenderField = {
  field: Field,
  flexFieldLayout: FlexFieldLayout,
  fieldValue?: any,
  components: FormComponents,
  props?: any
}

type RenderFieldControlled = {
  ...RenderField,
  handleInputChange: (event: SyntheticKeyboardEvent<EventTarget>) => void,
  handleInputFocus: (event: SyntheticKeyboardEvent<EventTarget>) => void,
  handleInputBlur: (event: SyntheticKeyboardEvent<EventTarget>) => void,
  components?: ?FormComponents
}

type RenderForm = FormProps & {
  formName: string,
  fields: Array<Field>,
  flexFieldLayout: FlexFieldLayout,
  flexLayout: FlexLayout,
  handleSubmit: Function,
  errors?: Errors,
  showErrors?: boolean,
  renderFieldMethod: Function,
  renderButtonsMethod: Function,
  buttonsProps?: RenderButtons,
  translations?: {
    reset?: string,
    submit?: string
  },
  components: FormComponents
}

class FormSimple extends PureComponent<FormSimpleProps> {

  static defaultProps: {
    flexLayout: FlexLayout,
    flexFieldLayout: FlexFieldLayout
  } = {
    flexFieldLayout: {
      minWidth: '25%',
      maxWidth: '25%'
    },
    flexLayout: {
      wrap: true
    }
  }

  constructor(props: FormSimpleProps) {
    super(props)

    this.memoizedHandleFieldInputChange = memoize(this.handleFieldInputChange, (field) => field.name)
    this.memoizedHandleFieldInputFocus = memoize(this.eventHandler('focus'), (field) => field.name)
    this.memoizedHandleFieldInputBlur = memoize(this.eventHandler('blur'), (field) => field.name)
  }

  handleSubmit = (event: SyntheticEvent<EventTarget>) => {
    event.preventDefault()
    this.props.onSubmitHandler()
  }

  handleFieldInputChange = (field: Field) => (event: SyntheticEvent<EventTarget>) => {
    if (event.target instanceof HTMLInputElement) {
      const { value }: { value: any } = event.target
      let newValue: any = value
      if (field.valueDecorator) {
        newValue = field.valueDecorator(value, field)
      }
      const { name }: { name: string } = field

      field.onChange && field.onChange(name, newValue)
      this.props.onChangeHandler && this.props.onChangeHandler(name, newValue)
    }
  }

  eventHandler = (eventName: string) => {
    const eventPreffix: string = `on${eventName.substr(0, 1).toUpperCase()}${eventName.substr(1)}`
    const eventPreffixHandler: string = `${eventPreffix}Handler`
    return (field: Field) => (event: SyntheticEvent<EventTarget>) => {
      if (event.target instanceof HTMLInputElement) {
        const { value }: { value: any }  = event.target
        const { name }: { name: string } = field
        field[eventPreffix] && field[eventPreffix](name, value)
        this.props[eventPreffixHandler] && this.props[eventPreffixHandler](name, value)
      }
    }
  }

  resetHandler = () => {
    this.props.onResetHandler && this.props.onResetHandler()
  }

  memoizedHandleFieldInputChange: Function
  memoizedHandleFieldInputFocus: Function
  memoizedHandleFieldInputBlur: Function
  props: FormSimpleProps

  renderButtons = ({ showReset, submitDisabled, resetable,
                     labels: { reset, submit }, errors, components }: RenderButtons) => {
    const hasErrors = !!(errors && (errors.length || Object.values(errors).length))
    const buttonsData: Array<Object> = []
    if (showReset) {
      buttonsData.push({ big: true, type: 'reset', disabled: !resetable, onClick: this.resetHandler, label: reset })
    }
    buttonsData.push({ big: true,
      isPlatinum: true,
      type: 'submit',
      disabled: submitDisabled || hasErrors,
      label: submit
    })

    return (this.props.renderButtons
      ? this.props.renderButtons(buttonsData)
      : <components.ActionGroupComponent data={buttonsData} />)
  }

  renderForm = (formProps: FormProps) =>
    (this.props.renderForm ? this.props.renderForm(formProps) : <formProps.components.FormComponent {...formProps} />)

  renderField = ({ field, flexFieldLayout, fieldValue, components, ...props }: RenderField) => {
    const handleInputChange: Function = this.memoizedHandleFieldInputChange(field)
    const handleInputFocus: Function = this.memoizedHandleFieldInputFocus(field)
    const handleInputBlur: Function = this.memoizedHandleFieldInputBlur(field)

    // TODO FORM Field props type
    const formFieldProps: RenderFieldControlled = {
      field, flexFieldLayout, fieldValue, handleInputChange, handleInputFocus, handleInputBlur, ...props }

    const renderFieldMethod = field.render || this.props.renderField
    return renderFieldMethod ? renderFieldMethod(formFieldProps)
      : <components.FormFieldComponent key={field.name} {...formFieldProps} />
  }

  render() {
    const {
      fields, errors, showErrors, formName, buttonLabels,
      flexLayout, flexFieldLayout, showReset, submitDisabled, translations, resetable, components,
      ...formProps
    }: FormSimpleProps = this.props
    const buttonsProps: RenderButtons = {
      showReset,
      submitDisabled,
      resetable,
      labels: buttonLabels || {
          reset: translations && translations.reset ? translations.reset : 'Reset',
          submit: translations && translations.submit ? translations.submit : 'Search'
      },
      components,
      errors
    }
    const renderFormProps: RenderForm = {
      formName,
      fields,
      flexFieldLayout,
      flexLayout,
      handleSubmit: this.handleSubmit,
      errors,
      showErrors,
      renderFieldMethod: this.renderField,
      renderButtonsMethod: this.renderButtons,
      buttonsProps,
      translations,
      components,
      ...formProps
    }

    return this.renderForm(renderFormProps)
  }
}
export default FormSimple
