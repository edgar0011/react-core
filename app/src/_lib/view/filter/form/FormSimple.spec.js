import React from 'react'
import FormSimple from './FormSimple'
import Form from './Form'
import FormField from './FormField'
import ActionGroup from '../../button/ActionGroup'

describe('FormSimple', () => {
  it('should render FormSimple', () => {
    const fields = [
      {
        name: 'input1',
        label: 'Input1'
      }
    ]
    const fieldValues = {
      input1: 'Hello Form'
    }

    const formComponents = {
      FormComponent: Form,
      FormFieldComponent: FormField,
      ActionGroupComponent: ActionGroup
    }

    const translations = {}

    global.shallowSnapshot(
      <FormSimple
        fields={fields}
        fieldValues={fieldValues}
        translations={translations}
        components={formComponents}
      />)
  })
})
