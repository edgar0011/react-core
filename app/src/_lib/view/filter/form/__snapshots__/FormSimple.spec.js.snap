// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormSimple should render FormSimple 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <FormSimple
    components={
      Object {
        "ActionGroupComponent": [Function],
        "FormComponent": [Function],
        "FormFieldComponent": [Function],
      }
    }
    fieldValues={
      Object {
        "input1": "Hello Form",
      }
    }
    fields={
      Array [
        Object {
          "label": "Input1",
          "name": "input1",
        },
      ]
    }
    flexFieldLayout={
      Object {
        "maxWidth": "25%",
        "minWidth": "25%",
      }
    }
    flexLayout={
      Object {
        "wrap": true,
      }
    }
    translations={Object {}}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "buttonsProps": Object {
        "components": Object {
          "ActionGroupComponent": [Function],
          "FormComponent": [Function],
          "FormFieldComponent": [Function],
        },
        "errors": undefined,
        "labels": Object {
          "reset": "Reset",
          "submit": "Search",
        },
        "resetable": undefined,
        "showReset": undefined,
        "submitDisabled": undefined,
      },
      "components": Object {
        "ActionGroupComponent": [Function],
        "FormComponent": [Function],
        "FormFieldComponent": [Function],
      },
      "errors": undefined,
      "fieldValues": Object {
        "input1": "Hello Form",
      },
      "fields": Array [
        Object {
          "label": "Input1",
          "name": "input1",
        },
      ],
      "flexFieldLayout": Object {
        "maxWidth": "25%",
        "minWidth": "25%",
      },
      "flexLayout": Object {
        "wrap": true,
      },
      "formName": undefined,
      "handleSubmit": [Function],
      "renderButtonsMethod": [Function],
      "renderFieldMethod": [Function],
      "showErrors": undefined,
      "translations": Object {},
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "buttonsProps": Object {
          "components": Object {
            "ActionGroupComponent": [Function],
            "FormComponent": [Function],
            "FormFieldComponent": [Function],
          },
          "errors": undefined,
          "labels": Object {
            "reset": "Reset",
            "submit": "Search",
          },
          "resetable": undefined,
          "showReset": undefined,
          "submitDisabled": undefined,
        },
        "components": Object {
          "ActionGroupComponent": [Function],
          "FormComponent": [Function],
          "FormFieldComponent": [Function],
        },
        "errors": undefined,
        "fieldValues": Object {
          "input1": "Hello Form",
        },
        "fields": Array [
          Object {
            "label": "Input1",
            "name": "input1",
          },
        ],
        "flexFieldLayout": Object {
          "maxWidth": "25%",
          "minWidth": "25%",
        },
        "flexLayout": Object {
          "wrap": true,
        },
        "formName": undefined,
        "handleSubmit": [Function],
        "renderButtonsMethod": [Function],
        "renderFieldMethod": [Function],
        "showErrors": undefined,
        "translations": Object {},
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
